pkg/util/names.go

```diff
@@ -0,0 +1,29 @@
+package util
+
+import (
+	"crypto/sha256"
+	"fmt"
+	"regexp"
+)
+
+var validNamePattern = regexp.MustCompile(`^[a-z][a-z0-9\-]*$`).MatchString
+
+// IsValidCK8sCRDName checks if a name can be used for a kubernetes CRD resource
+// * contain at most 63 characters
+// * start with a lowercase alpha character
+// * contain only lowercase alphanumeric characters or '-'
+func IsValidCK8sCRDName(name string) bool {
+	return validNamePattern(name) && len(name) < 64
+}
+
+// GrafanaUIDToK8sName defines a consistent mapping between Grafana UIDs and k8s compatible names
+func GrafanaUIDToK8sName(uid string) string {
+	if IsValidCK8sCRDName(uid) {
+		return uid // OK, so just use it directly
+	}
+
+	h := sha256.New()
+	_, _ = h.Write([]byte(uid))
+	bs := h.Sum(nil)
+	return fmt.Sprintf("g%x", bs[:12])
```

Comment:
The biggest issue I have with this kind of approach is that there's no way to reverse the hash and get back to the raw uid.  I'd be tempted to use something like https://pkg.go.dev/encoding/base32 which we could convert back.  That doesn't help us if it has more than 63 characters after encoding though...

If we don't have reversible k8s names then we end up needing to store them as well to be able to resolve a name

Reply:
yup -- I don't know how to get around it though.

That said, i have long suggested we should support "alias" ids that can redirect old ids to the new ones